### PR TITLE
NMS-14215: Add encryption option for snmp config 

### PR DESCRIPTION
**Pull request by cgorantla**
### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-14215

### CODE REVIEW

**Comment by cgorantla** [comment]
This is in Draft as I am still validating more scenarios but look good so far.

**Comment by christianpape** [comment]
Looks good so far. Only the headers of the new files need to point to year 2022.

File: docs/modules/operation/pages/provisioning/snmp-encryption.adoc [hunk 1]
```diff
@@ -0,0 +1,43 @@
+== Encrypting SNMP Credentials
```
Comment (Bonrob2) [inline docs/modules/operation/pages/provisioning/snmp-encryption.adoc:1]:
```suggestion
== Encrypt SNMP credentials
```

File: docs/modules/operation/pages/provisioning/snmp-encryption.adoc [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+You can save SNMP credentials like readCommunity/WriteCommunity, authPassPhrase/privPassPhrase on the file system in encrypted format.
+
+=== Encryption methodology
+
+A random password is generated (once) for `snmp-config` alias and it is stored in java keystore file at `$OPENNMS_HOME/etc/scv.jce`.
+This password is used to encrypt/decrypt SNMP credentials.
+
+=== Enabling Encrption on SNMP Config
+
+Enable encryption on SNMP config by setting the following system property:
```
Comment (Bonrob2) [inline docs/modules/operation/pages/provisioning/snmp-encryption.adoc:12]:
```suggestion
Enable encryption on SNMP configuration by setting the following system property:
```

File: docs/modules/operation/pages/provisioning/snmp-encryption.adoc [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+[source, sh]
+----
+echo 'org.opennms.snmp.encryption.enabled=true' >> "$OPENNMS_HOME/etc/opennms.properties.d/snmp-encryption.properties"
+----
+
+SNMP config that is modified through the user interface will be encrypted directly on the file system.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/provisioning/snmp-encryption.adoc:18]:
```suggestion
SNMP configuration modified through the user interface is encrypted directly on the file system.
```
I'm not clear if this is a comment about how the UI functions, or this is describing a result of what happens with anything that has been modified via the UI as a result of running this command.

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
A few changes. Thanks!

File: core/snmp/commands/src/main/java/org/opennms/netmgt/snmp/commands/EncryptSnmpConfigCommand.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2020 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2020 The OpenNMS Group, Inc.
```
Comment (christianpape) [inline core/snmp/commands/src/main/java/org/opennms/netmgt/snmp/commands/EncryptSnmpConfigCommand.java:5]:
```suggestion
 * Copyright (C) 2022 The OpenNMS Group, Inc.
 * OpenNMS(R) is Copyright (C) 1999-2022 The OpenNMS Group, Inc.
```

File: core/snmp/commands/src/main/java/org/opennms/netmgt/snmp/commands/EncryptSnmpConfigCommand.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Reference
+    public SnmpAgentConfigFactory snmpAgentConfigFactory;
+
+    @Override
+    public Object execute() throws Exception {
+        boolean encryptionEnabled = Boolean.getBoolean("org.opennms.snmp.encryption.enabled");
+        if (!encryptionEnabled) {
+            System.out.println("Encryption is not enabled, \n" +
+                    " Set System property `org.opennms.snmp.encryption.enabled` to true to enable encryption on Snmp.");
```
Comment (christianpape) [inline core/snmp/commands/src/main/java/org/opennms/netmgt/snmp/commands/EncryptSnmpConfigCommand.java:49]:
```suggestion
                    " Set system property `org.opennms.snmp.encryption.enabled` to true to enable encryption on Snmp.");
```

File: core/snmp/commands/src/main/java/org/opennms/netmgt/snmp/commands/DecryptSnmpConfigCommand.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    @Option(name = "-f", aliases = "--filename", description = "File name where decrypted SNMP")
+    String fileName = "snmp-config-decrypt.xml";
+
+    @Override
+    public Object execute() throws Exception {
+
+        boolean encryptionEnabled = Boolean.getBoolean("org.opennms.snmp.encryption.enabled");
+        if (!encryptionEnabled) {
+            System.out.println("Encryption is not enabled, \n" +
+                    " Set System property `org.opennms.snmp.encryption.enabled` to true to enable encryption on Snmp.");
```
Comment (christianpape) [inline core/snmp/commands/src/main/java/org/opennms/netmgt/snmp/commands/DecryptSnmpConfigCommand.java:64]:
```suggestion
                    " Set system property `org.opennms.snmp.encryption.enabled` to true to enable encryption on Snmp.");
```

File: docs/modules/operation/pages/provisioning/snmp-encryption.adoc [hunk 7]
```diff
@@ -0,0 +1,43 @@
+== Encrypting SNMP Credentials
+
+You can save SNMP credentials like readCommunity/WriteCommunity, authPassPhrase/privPassPhrase on the file system in encrypted format.
```
Comment (christianpape) [inline docs/modules/operation/pages/provisioning/snmp-encryption.adoc:3]:
```suggestion
You can save SNMP credentials like readCommunity/writeCommunity, authPassPhrase/privPassPhrase on the file system in encrypted format.
```

File: docs/modules/operation/pages/provisioning/snmp-encryption.adoc (hunk 3 above)
Comment (christianpape) [inline docs/modules/operation/pages/provisioning/snmp-encryption.adoc:18]:
```suggestion
SNMP config that is modified through the user interface will be encrypted directly on the filesystem.
```

**Review by christianpape — CHANGES_REQUESTED** [review]
Only a few minor changes needed.

File: docs/modules/operation/pages/provisioning/snmp-encryption.adoc (hunk 3 above)
Comment (cgorantla) [reply docs/modules/operation/pages/provisioning/snmp-encryption.adoc]: we are saying SNMP configuration that is modified on UI will get encrypted immediately on the file system.

Comment (cgorantla) [reply docs/modules/operation/pages/provisioning/snmp-encryption.adoc]: > > we are saying SNMP configuration that is modified on UI will get encrypted immediately on the file system.
> 
> Thanks. I updated my edit slightly.

Handled.

**Review by christianpape — APPROVED** [review]
LGTM.

**Review by Bonrob2 — APPROVED** [review]
LGTM!

**Pull request merged**